lib/p2p/Peer.ts

```diff
@@ -1,20 +1,251 @@
-function Peer({
-    host, address, port,
-  }) {
-  this.host = host;
-  this.address = address;
-  this.port = port;
+import assert from 'assert';
+import net, { Socket } from 'net';
+import { EventEmitter } from 'events';
+import NetAddress from './NetAddress';
+import Parser, { ParserError, ParserErrorType } from './Parser';
+import Logger from '../Logger';
+
+const pubKey = `tempPK_${Math.floor(1000 + (Math.random() * 9000))}`;
+
+enum ConnectionDirection {
+  INBOUND,
+  OUTBOUND,
 }
 
-Peer.prototype.getKey = function getKey() {
-  return `${this.address}:${this.port}`;
-};
+/** Represents a remote XU peer */
+class Peer extends EventEmitter {
+  public address!: NetAddress;
+  private logger: Logger;
+  private connected: boolean;
+  private opened: boolean;
+  private direction?: ConnectionDirection;
+  private socket!: Socket;
+  private parser: Parser;
+  private destroyed: boolean;
+  private connectTimeout?: NodeJS.Timer;
+  private connectTime: number;
+  private banScore: number;
+  private lastRecv: number;
+  private lastSend: number;
+
+  get id(): string {
+    assert(this.address);
+    return this.address.hostname;
+  }
+
+  get statusString(): string {
+    if (this.connected) {
+      return `Connected to peer (${this.id})`;
+    } else {
+      return 'Not connected';
+    }
+  }
+
+  constructor() {
+    super();
+
+    this.logger = Logger.global;
+    this.connected = false;
+    this.opened = false;
+    this.parser = new Parser();
+    this.destroyed = false;
+    this.connectTime = 0;
+    this.banScore = 0;
+    this.lastRecv = 0;
+    this.lastSend = 0;
+
+    this._init();
+  }
+
+  static fromOutbound(address: NetAddress): Peer {
+    const peer = new Peer();
+    peer.connect(address);
+    return peer;
+  }
+
+  static fromInbound(socket: Socket): Peer {
+    const peer = new Peer();
+    peer.accept(socket);
+    return peer;
+  }
+
+  public open = async (): Promise<void> => {
+    this.opened = true;
```

Comment:
In principle, `open` represent the full initialization process, which starts with opening the connection. The distinction would more clear when we’ll implement the handshake. 
The idea is that after opening the connection, we’ll want to send some data to the other side. The receiver side might not enlist the peer in the pool until he will get that data. That data can be things like like your public key, xud version, incoming socket port, accepted currency pairs, etc. 
in regards to `opened` property - I haven’t finished implementing that functionality yet. I thought to do so after implementing some of the handshake steps.

Reply:
OK thanks for clarifying. So `opened` essentially means a socket with a peer is ready for use by xud. Maybe `ready` is a better name, perhaps `initialized`? Also do you agree that the `initConnection` and `open` methods can be collapsed into one, or is there a case where you would envision `initConnection` being called somewhere besides `open`? I prefer the `initConnection` name, especially if `opened` is renamed.

Reply:
`open` and `initConnection` shouldn't get merged, even though `initConnection` might not be called from anywhere else, because it has a specific and distinct responsibility, and a bunch of code which I wouldn't want to see on `open`.
There's gonna be a few async methods, similar to `initConnection`, which will constructs the process. it's gonna block the peer from being added to the Pool's peers list, until it will be completed.

the `opened` state doesn't mean anything fancy for now. just that process of connecting + handshaking with the peer has began, and you shouldn't attempt to start it again. i'm gonna add that small logic in the next commits.